secretsclient/client.go

```diff
@@ -0,0 +1,388 @@
+package secretsclient
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+
+	"github.com/earthly/earthly/secretsclient/api"
+
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/golang/protobuf/proto"
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/ssh/agent"
+)
+
+// ErrAccountExists occurs account creation when an account already exists
+var ErrAccountExists = fmt.Errorf("account already exists")
+
+// ErrUnauthorized occurs when a user is unauthorized to access a resource
+var ErrUnauthorized = fmt.Errorf("unauthorized")
+
+// Client provides a client to the shared secrets service
+type Client interface {
+	RegisterEmail(email string) error
+	CreateAccount(email, verificationToken, password, publicKey string) error
+	Get(path string) ([]byte, error)
+	Set(path string, data []byte) error
+	GetPublicKeys() ([]*agent.Key, error)
+	CreateOrg(org string) error
+}
+
+type request struct {
+	hasBody bool
+	body    []byte
+
+	hasAuth bool
+}
+type requestOpt func(*request) error
+
+func withPublicKeyAuth() requestOpt {
+	return func(r *request) error {
+		r.hasAuth = true
+		return nil
+	}
+}
+
+func withJSONBody(body proto.Message) requestOpt {
+	return func(r *request) error {
+		marshaler := jsonpb.Marshaler{}
+		encodedBody, err := marshaler.MarshalToString(body)
+		if err != nil {
+			return err
+		}
+
+		r.hasBody = true
+		r.body = []byte(encodedBody)
+		return nil
+	}
+}
+
+func withBody(body string) requestOpt {
+	return func(r *request) error {
+		r.hasBody = true
+		r.body = []byte(body)
+		return nil
+	}
+}
+
+func (c *client) doCall(method, url string, opts ...requestOpt) (int, string, error) {
+	var r request
+	for _, opt := range opts {
+		err := opt(&r)
+		if err != nil {
+			return 0, "", err
+		}
+	}
+	var bodyReader io.Reader
+	var bodyLen int64
+	if r.hasBody {
+		bodyReader = bytes.NewReader(r.body)
+		bodyLen = int64(len(r.body))
+	}
+
+	req, err := http.NewRequest(method, c.secretServer+url, bodyReader)
+	if err != nil {
+		return 0, "", err
+	}
+	if bodyReader != nil {
+		req.ContentLength = bodyLen
+	}
+	if r.hasAuth {
+		_, authToken, err := c.getAuthToken()
+		if err != nil {
+			return 0, "", err
+		}
+		req.Header.Add("Authorization", authToken)
+	}
+
+	client := &http.Client{}
+
+	resp, err := client.Do(req)
```

Comment:
We should probably bake in some auto-retry logic. We'll probably have to be very careful with `4xx` vs `5xx` status codes on the server side.

In any case, this could be a `TODO` item.

Reply:
that's a good point. i'll make a TODO.